src/main/java/com/tananaev/adblib/AdbConnection.java

```diff
@@ -271,12 +286,16 @@ public void connect() throws IOException, InterruptedException {
         /* Wait for the connection to go live */
         synchronized (this) {
             if (!connected)
-                wait();
+                wait(unit.toMillis(timeout));
 
-            if (!connected) {
-                throw new IOException("Connection failed");
-            }
+            if (!connected)
+                if (connectAttempted)
+                    return false;
```

Comment:
Why not an exception?

Reply:
> Why not an exception?

My first preference was to throw a `java.util.concurrent.TimeoutException`, but [the Javadoc for that class](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/TimeoutException.html) says you should use a return value instead:

> Exception thrown when a blocking operation times out. Blocking operations for which a timeout is specified need a means to indicate that the timeout has occurred. For many such operations it is possible to return a value that indicates timeout; when that is not possible or desirable then TimeoutException should be declared and thrown.

I then checked a few spots in Java where timeouts are used, and I noticed that none of them throw an exception on Timeout:
* [`ReentrantLock.tryLock()`](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReentrantLock.html#tryLock(long,%20java.util.concurrent.TimeUnit))
* [`CountDownLatch.await()`](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CountDownLatch.html#await(long,%20java.util.concurrent.TimeUnit))
* [`Object.wait()`](https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#wait(long))

So I thought I'd better stick with the convention.

On the other hand, one benefit of throwing an exception is we could extend the timeout functionality to also apply to the connection attempts inside `AdbConnection.getMaxData()` and `AdbConnection.open()`

Personally, I'm happy to go with either a return value or an exception. Would you like me to swap it over to an exception?

Reply:
Sounds reasonable. Let's merge this part.

Reply:
Thanks!